### PR TITLE
Optimize libnode's annotation/label modification functions + stabilize & fix migration functests

### PR DESCRIPTION
**Pull request by iholder-redhat**
**What this PR does / why we need it**:
Currently, libnode marshalls the whole node just to add a label/annotation. Then, it patches the whole node.
This is clearly wasteful, and with such a basic operation that could slow up our whole testing suite.

This PR avoids marshalling and patches only the labels / annotations in an efficient way. In addition, it adds support for label/annotation removal.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: tests/infra_test.go [hunk 1]
```diff
@@ -1180,29 +1180,8 @@ var _ = Describe("[Serial][sig-compute]Infrastructure", func() {
...
-					Value: node.Annotations,
-				})
-
-				payloadBytes, err := json.Marshal(p)
-				Expect(err).ToNot(HaveOccurred())
-
-				_, err = virtClient.CoreV1().Nodes().Patch(context.Background(), node.Name, types.JSONPatchType, payloadBytes, metav1.PatchOptions{})
-				Expect(err).ToNot(HaveOccurred())
+				libnode.RemoveLabelFromNode(node.Name, nonExistingCPUModelLabel)
+				libnode.RemoveAnnotationFromNode(node.Name, v1.LabellerSkipNodeAnnotation)
```
Comment (brybacki) [inline tests/infra_test.go:1184]: looks so much better

File: tests/libnode/node.go [hunk 2]
```diff
@@ -122,56 +123,80 @@ func GetNodeDrainKey() string {
 	return virtconfig.NodeDrainTaintDefaultKey
 }
 
-func addLabelAnnotationHelper(nodeName, key, value string, isLabel bool) {
+// Adds or removes a label or annotation from a node. When removing a label/annotation, the "value" parameter
+// is ignored.
+func addRemoveLabelAnnotationHelper(nodeName, key, value string, isLabel, toAdd bool) {
```
Comment (brybacki) [inline tests/libnode/node.go:128]: honestly, this is still not nice

File: tests/infra_test.go (hunk 1 above)
Comment (brybacki) [reply tests/infra_test.go:1184]: and probably works better

Comment (iholder-redhat) [reply tests/infra_test.go:1184]: Thanks!

File: tests/libnode/node.go (hunk 2 above)
Comment (iholder-redhat) [reply tests/libnode/node.go:128]: The names are tricky here. Do you have any ideas?

Comment (iholder-redhat) [reply tests/libnode/node.go]: PTAL.
Now the caller would need to provide a `mapType` (which is basically labels or annotations) and an `actionType` (which is add or remove). This is declarative, explicit, readable and safe.

The helper function itself first assigns the right functions to fetch a map, mutate the map, etc. Afterwards, it uses these functions to perform its logic in a transparent way from a readability point of view.

WDYT?

File: tests/libnode/node.go [hunk 3]
```diff
@@ -122,56 +123,94 @@ func GetNodeDrainKey() string {
...
+type mapAction string
+
+const (
+	add    mapAction = "add"
+	remove mapAction = "remove"
+)
+
+// Adds or removes a label or annotation from a node. When removing a label/annotation, the "value" parameter
+// is ignored.
+func addRemoveLabelAnnotationHelper(nodeName, key, value string, mapType mapType, mapAction mapAction) {
```
Comment (brybacki) [inline tests/libnode/node.go:142]:
```suggestion
it is a stepo into the right direction, WDYT about this?

func patchNodeMapHelper(nodeName string, type mapType, 
fetchMap func(node *k8sv1.Node) map[string]string,
mutateMap func(key, val string, m map[string]string) map[string]string) {

```
and use it like this:
```
fetchLabelsFunc = func(node *k8sv1.Node) map[string]string { return node.Labels }
fetchAnnotationsFunc = func(node *k8sv1.Node) map[string]string { return node.Annotations }

addKvFunc = func(key, value string)  	func(m map[string]string) map[string]string  {
return func(m map[string]string) map[string]string {
			m[key] = val
			return m
		}
}

func AddLabelToNode(nodeName, key, value string) {
	addRemoveLabelAnnotationHelper(nodeName, label,
	fetchLabels, 
	addKvFunc(key, value))
}

func AddAnnotationToNode(nodeName, key, value string) {
	addRemoveLabelAnnotationHelper(nodeName, annotation
	fetchAnnotationsFunc, 
	addKvFunc(key, value))
}
```

Comment (brybacki) [reply tests/libnode/node.go:142]: It is a pity that `go` does not have nice functional syntax sugar for anonymous functions

Comment (iholder-redhat) [reply tests/libnode/node.go:142]: I thought about doing it in the way you described, but I'm not sure it is better. My thinking was that these function are exclusively related to `addRemoveLabelAnnotationHelper()` function, so why expose them into the global outer scope?

I think this way is better since it's obvious what's the purpose of these functions and the scope which they serve.

Comment (iholder-redhat) [reply tests/libnode/node.go:142]: > It is a pity that go does not have nice functional syntax sugar for anonymous functions

Indeed!!

**Comment by iholder-redhat** [comment]
BTW @brybacki, the tests all pass locally and don't seem to even use the functions introduced in this PR.. Do you have an idea what could be wrong here?

**Comment by brybacki** [comment]
> BTW @brybacki, the tests all pass locally and don't seem to even use the functions introduced in this PR.. Do you have an idea what could be wrong here?
Do you have multiple nodes on local cluster?

**Comment by iholder-redhat** [comment]
> > BTW @brybacki, the tests all pass locally and don't seem to even use the functions introduced in this PR.. Do you have an idea what could be wrong here?
> Do you have multiple nodes on local cluster?

Yes, I do

**Comment by iholder-redhat** [comment]
The odd thing here is that all infra_tests are serial, and none of them use the functions I've introduced in this PR :thinking:

**Comment by Barakmor1** [comment]
Hey ,
Thanks for that actually there are flacky tests that fail because we update the origanal node labels\annotations regading the node labeller instead of using this logic of Patch.
Can you please use the new logic in:
`Context("[Serial]Should trigger event if vmi with host-model start on source node with uniq host-model", func() `
`Context("[Serial]Should trigger event if the nodes doesn't contain MigrationSelectorLabel for the vmi host-model type", func()`
`Context("[Serial]Testing host-model cpuModel edge cases in the cluster if the cluster is host-model migratable", func()`
I think that we should add enbleNodeLabller function (new function) just for removing the annotation that skip node labeller instead of messing with all the node's labels and annotation , the node labeller will do the rest .

File: tests/libnode/node.go [hunk 4]
```diff
@@ -122,56 +123,94 @@ func GetNodeDrainKey() string {
...
-
-	patch, err := strategicpatch.CreateTwoWayMergePatch(old, newJson, node)
-	Expect(err).ToNot(HaveOccurred())
+	addRemoveLabelAnnotationHelper(nodeName, key, "", label, remove)
+}
 
-	_, err = virtCli.CoreV1().Nodes().Patch(context.Background(), node.Name, types.StrategicMergePatchType, patch, k8smetav1.PatchOptions{})
-	Expect(err).ToNot(HaveOccurred())
+func RemoveAnnotationFromNode(nodeName string, key string) {
+	addRemoveLabelAnnotationHelper(nodeName, key, "", label, remove)
```
Comment (Barakmor1) [inline tests/libnode/node.go:213]: should be `addRemoveLabelAnnotationHelper(nodeName, key, "", annotation, remove)`

**Comment by iholder-redhat** [comment]
/hold
This PR is blocked by https://github.com/kubevirt/kubevirt/pull/7872

File: tests/migration_test.go [hunk 5]
```diff
@@ -4194,25 +4136,97 @@ func temporaryTLSConfig() *tls.Config {
...
 	key, value := v1.LabellerSkipNodeAnnotation, "true"
-	libnode.AddAnnotationToNode(node.Name, key, value)
+	libnode.AddAnnotationToNode(nodeName, key, value)
 
-	By(fmt.Sprintf("Expecting node %s to include %s label", node.Name, v1.LabellerSkipNodeAnnotation))
+	By(fmt.Sprintf("Expecting node %s to include %s label", nodeName, v1.LabellerSkipNodeAnnotation))
 	Eventually(func() bool {
-		node, err = virtClient.CoreV1().Nodes().Get(context.Background(), node.Name, metav1.GetOptions{})
+		node, err = virtClient.CoreV1().Nodes().Get(context.Background(), nodeName, metav1.GetOptions{})
+		Expect(err).ShouldNot(HaveOccurred())
```
Comment (jean-edouard) [inline tests/migration_test.go:4153]: Nit:
```suggestion
		Expect(err).ToNot(HaveOccurred())
```

File: tests/migration_test.go [hunk 6]
```diff
@@ -4194,25 +4136,97 @@ func temporaryTLSConfig() *tls.Config {
...
+
+	return node
+}
+
+func resumeNodeLabeller(nodeName string, virtClient kubecli.KubevirtClient) *k8sv1.Node {
+	var err error
+	var node *k8sv1.Node
+
+	node, err = virtClient.CoreV1().Nodes().Get(context.Background(), nodeName, metav1.GetOptions{})
+	Expect(err).ShouldNot(HaveOccurred())
```
Comment (jean-edouard) [inline tests/migration_test.go:4167]: Nit:
```suggestion
	Expect(err).ToNot(HaveOccurred())
```

File: tests/migration_test.go [hunk 7]
```diff
@@ -4194,25 +4136,97 @@ func temporaryTLSConfig() *tls.Config {
...
+			return fmt.Errorf("node %s is expected to have a label with %s prefix. this means node-labeller is not enabled for the node", nodeName, v1.HostModelCPULabel)
+		}
+
+		return nil
+	}, 30*time.Second, time.Second).ShouldNot(HaveOccurred())
 
 	return node
 }
 
+func wakeNodeLabellerUp(virtClient kubecli.KubevirtClient) {
```
Comment (jean-edouard) [inline tests/migration_test.go:4216]: SuperNit:
```suggestion
func wakeUpNodeLabeller(virtClient kubecli.KubevirtClient) {
```

**Review by jean-edouard — COMMENTED** [review]
/approve
Couple nits in-line, feel free to ignore

File: tests/migration_test.go [hunk 8]
```diff
@@ -2784,58 +2759,37 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
...
-						var currNode *k8sv1.Node
-						Eventually(func() map[string]string {
-							currNode, err = virtClient.CoreV1().Nodes().Get(context.Background(), node.Name, metav1.GetOptions{})
-							Expect(err).ShouldNot(HaveOccurred())
-							return currNode.Labels
-						}, 10*time.Second, 1*time.Second).Should(Equal(originalNodeLabels[indx]), "Node should have original labels")
-						Expect(currNode.Annotations).To(Equal(originalNodeAnnotations[indx]))
+					for _, node := range nodes {
+						resumeNodeLabeller(node.Name, virtClient)
 					}
```
Comment (Barakmor1) [inline tests/migration_test.go:2801]: Can you verify that labels with `v1.SupportedHostModelMigrationCPU` are in the node?

File: tests/migration_test.go [hunk 9]
```diff
@@ -2719,26 +2708,12 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
...
-
-					Eventually(func() map[string]string {
-						node, err = virtClient.CoreV1().Nodes().Get(context.Background(), node.Name, metav1.GetOptions{})
-						Expect(err).ShouldNot(HaveOccurred())
-
-						return node.Labels
-					}, 10*time.Second, 1*time.Second).Should(Equal(originalNodeLabels), "Node should have fake host model")
-					Expect(node.Annotations).To(Equal(originalNodeAnnotations))
+					By("Resuming node labeller")
+					node = resumeNodeLabeller(node.Name, virtClient)
```
Comment (Barakmor1) [inline tests/migration_test.go:2738]: please verify that `node.Labels[v1.HostModelCPULabel+"fake-model"]` doesn't exist.

Actually i think it exist after the test because in:

https://github.com/kubevirt/kubevirt/blob/970d68ae81247e082c708d54fbf14cfb70c728bb/pkg/virt-handler/node-labeller/node_labeller.go#L305

the labeller never delete it ,this might cause flackiness in later tests and it will be VERY hard to catch this.

**Comment by Barakmor1** [comment]
Can you please add:
```
			strings.Contains(label, kubevirtv1.HostModelCPULabel) ||
			strings.Contains(label, kubevirtv1.HostModelRequiredFeaturesLabel) ||
```
to https://github.com/kubevirt/kubevirt/blob/970d68ae81247e082c708d54fbf14cfb70c728bb/pkg/virt-handler/node-labeller/node_labeller.go#L305

File: tests/migration_test.go (hunk 6 above)
Comment (iholder-redhat) [reply tests/migration_test.go:4167]: Done

File: tests/migration_test.go (hunk 5 above)
Comment (iholder-redhat) [reply tests/migration_test.go:4153]: Done

File: tests/migration_test.go (hunk 7 above)
Comment (jean-edouard) [reply tests/migration_test.go:4216]: Yes it is, hence the "SuperNit" :) It's just slightly harder to read and slightly frowned upon. (see what I did there? :smile: )

File: tests/migration_test.go [hunk 10]
```diff
@@ -2784,58 +2759,37 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
...
-				var originalNodeAnnotations []map[string]string
 				var vmi *v1.VirtualMachineInstance
-				var backupNodes []k8sv1.Node
+				var nodes []k8sv1.Node
 
 				BeforeEach(func() {
-					backupNodes = libnode.GetAllSchedulableNodes(virtClient).Items
-					if len(backupNodes) == 1 || len(backupNodes) > 10 {
-						Skip("This test can't run with single node and it's too slow to run with more than 10 nodes")
-					}
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply tests/migration_test.go:2796]: Nice catch! It was removed by mistake. Brought it back :+1:

File: tests/migration_test.go [hunk 11]
```diff
@@ -3710,32 +3661,16 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
...
-
-			Eventually(func() map[string]string {
-				backedUpTargetNode, err = virtClient.CoreV1().Nodes().Get(context.Background(), targetNode.Name, metav1.GetOptions{})
-				Expect(err).ShouldNot(HaveOccurred())
-
-				return backedUpTargetNode.Labels
-			}, 10*time.Second, 1*time.Second).Should(Equal(originalTargetNodeLabels), "Node should not have amazingFeature after the test")
-
+			By("Resuming node labeller")
+			targetNode = resumeNodeLabeller(targetNode.Name, virtClient)
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply tests/migration_test.go:3676]: Awesome catch!! :superhero: 
Thanks a lot, also the checks you asked for are now added

File: tests/migration_test.go (hunk 9 above)
Comment (iholder-redhat) [reply tests/migration_test.go:2738]: Done. Thanks!!

File: tests/migration_test.go [hunk 12]
```diff
@@ -575,6 +575,28 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
...
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Expecting alert to be removed")
+		Eventually(func() []k8sv1.Event {
+			events, err := virtClient.CoreV1().Events(util.NamespaceTestDefault).List(context.Background(), eventListOpts)
+			Expect(err).ToNot(HaveOccurred())
+
+			return events.Items
+		}, 30*time.Second, 1*time.Second).Should(BeEmpty())
+	}
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply tests/migration_test.go:598]: Done

**Comment by Barakmor1** [comment]
Thanks for this PR i think that it will stable some flacky tests! :)
/lgtm

File: tests/migration_test.go [hunk 13]
```diff
@@ -4194,25 +4141,97 @@ func temporaryTLSConfig() *tls.Config {
...
+		Expect(err).ToNot(HaveOccurred())
+
+		value, exists := node.Annotations[v1.LabellerSkipNodeAnnotation]
+		return exists && value == "true"
+	}, 30*time.Second, time.Second).Should(BeTrue(), fmt.Sprintf("node %s is expected to have annotation %s", nodeName, v1.LabellerSkipNodeAnnotation))
+
+	return node
+}
+
+func resumeNodeLabeller(nodeName string, virtClient kubecli.KubevirtClient) *k8sv1.Node {
```
Comment (xpivarc) [inline tests/migration_test.go:4186]: I am wondering why we need to wait for the handler. Couldn't we just add labels that we remove/remove that we add and so on?

File: tests/infra_test.go [hunk 14]
```diff
@@ -1184,6 +1184,19 @@ var _ = Describe("[Serial][sig-compute]Infrastructure", func() {
 			}
 		})
 
+		expectNodeLabels := func(nodeName string, labelValidation func(map[string]string) (valid bool, errorMsg string)) {
+			var errorMsg string
+
+			Eventually(func() (isValid bool) {
```
Comment (xpivarc) [inline tests/infra_test.go:1190]: Please use offset

File: tests/migration_test.go [hunk 15]
```diff
@@ -2746,17 +2779,11 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
 					_ = tests.RunMigration(virtClient, migration)
 
 					By("Expecting for an alert to be triggered")
-					Eventually(func() []k8sv1.Event {
```
Comment (xpivarc) [inline tests/migration_test.go:2770]: Wouldn't it be enough to check it is attached to the VMI?

File: pkg/virt-handler/node-labeller/node_labeller.go [hunk 16]
```diff
@@ -313,7 +313,9 @@ func (n *NodeLabeller) removeLabellerLabels(node *v1.Node) {
 			strings.Contains(label, kubevirtv1.CPUTimerLabel) ||
 			strings.Contains(label, kubevirtv1.HypervLabel) ||
 			strings.Contains(label, kubevirtv1.RealtimeLabel) ||
-			strings.Contains(label, kubevirtv1.SEVLabel) {
```
Comment (xpivarc) [inline pkg/virt-handler/node-labeller/node_labeller.go:316]: I would probably make separate PR for this one.

**Review by xpivarc — COMMENTED** [review]
/hold
I think replacing the whole label or annotation field is not good. For example, consider node heartbeat.

File: tests/libnode/node.go [hunk 17]
```diff
@@ -122,56 +125,127 @@ func GetNodeDrainKey() string {
...
 
-	newJson, err := json.Marshal(new)
-	Expect(err).ToNot(HaveOccurred())
+	patchedNode := patchLabelAnnotationHelper(virtCli, nodeName, newMap, label)
+	Expect(equality.Semantic.DeepEqual(patchedNode.Labels, newMap)).To(BeTrue(), "patched node's labels are not as expected")
+	return patchedNode
+}
 
-	patch, err := strategicpatch.CreateTwoWayMergePatch(old, newJson, node)
+func PatchAnnotations(nodeName string, newMap map[string]string) *k8sv1.Node {
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply tests/libnode/node.go]: done

File: tests/libnode/node.go [hunk 18]
```diff
@@ -122,56 +125,127 @@ func GetNodeDrainKey() string {
...
+	p := []utiltype.PatchOperation{
+		{
+			Op:    "replace",
+			Path:  "/metadata/" + string(mapType) + "s",
+			Value: newMap,
+		},
+	}
+
+	patchBytes, err := json.Marshal(p)
+	Expect(err).ToNot(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply tests/libnode/node.go:152]: Why is it necessary here?

File: tests/migration_test.go (hunk 13 above)
Comment (iholder-redhat) [reply tests/migration_test.go:4186]: So we would ensure we have a clean testing environment.
This isn't different from what's being done with many functions in `BeforeTestCleanup()` for example.

As we have experienced before in these tests, backing up & restoring node labels manually is much more error prone. In addition, the overhead is negligible; we're talking about a matter of seconds for a very small number of tests.

File: tests/infra_test.go (hunk 14 above)
Comment (iholder-redhat) [reply tests/infra_test.go]: done

File: tests/migration_test.go (hunk 15 above)
Comment (iholder-redhat) [reply tests/migration_test.go:2770]: What do you mean exactly?

File: pkg/virt-handler/node-labeller/node_labeller.go (hunk 16 above)
Comment (iholder-redhat) [reply pkg/virt-handler/node-labeller/node_labeller.go]: There you go: https://github.com/kubevirt/kubevirt/pull/7928
I've assigned you to it, please take a look

**Comment by iholder-redhat** [comment]
> I think replacing the whole label or annotation field is not good. For example, consider node heartbeat.

replacing the whole label or annotation field? What do you mean by that? Can you be more specific and elaborate?

**Comment by Barakmor1** [comment]
> /hold I think replacing the whole label or annotation field is not good. For example, consider node heartbeat.

Nice observation !, we should avoid modifying node's metadata.

**Comment by iholder-redhat** [comment]
> /hold
> I think replacing the whole label or annotation field is not good. For example, consider node heartbeat.

Addressed, thanks for catching that.
/hold cancel
